### PR TITLE
Fix lzw decoding stopping short of full decode

### PR DESCRIPTION
**Pull request by HeroicKatora**
When the required amount of data was just a tiny bit larger than full a
page size (one buffer increment), the decoder would stop since all
compressed data had been read. However, the lzw decoder contains a small
internal bit buffer for code decoding. That is, some remaining bytes had
been retained and were still left to be decoded.

Analysis was complicated by a (probably) incorrect sign in a sanity
check: We had a check that the decompressed bytes were at most as large
as required. Instead, we need at least as much data as required while an
internal invariant should stop at the request size, but it's not a
critical format error. This has been replaced with more concise errors.

Closes: #106

### CODE REVIEW

File: src/decoder/mod.rs [hunk 1]
```diff
@@ -804,9 +804,12 @@ impl<R: Read + Seek> Decoder<R> {
             }
         };
 
-        if bytes / buffer.byte_len() > max_uncompressed_length {
+        if bytes / buffer.byte_len() < max_uncompressed_length {
```
Comment (HeroicKatora) [inline src/decoder/mod.rs:807]: I'm not 100% sure about this yet. It seems illogical to check if we had read _more_ data but currently we avoid this in `lzw` at least.

**Review by HeroicKatora — COMMENTED** [review]
Still looking for a small regression test. @ruffson @mntns
(The 600MB image passes though)

File: src/decoder/mod.rs (hunk 1 above)
Comment (HeroicKatora) [reply src/decoder/mod.rs:807]: This reveals an interesting test failure. `u12` data is packed in bytes but the decoder uses the `buffer.byte_len()` and multiplies it with the sample count to determine the expected buffer size. That's incorrect. Instead, the strip's data representation should be independent of the BitsPerSample.

Comment (fintelia) [reply src/decoder/mod.rs:807]: I think that the original code for this line was correct. Specifically for reasons I don't really understand, `max_uncompressed_length` is being used as the size for `buffer`. This check is to make sure that the subsequent indexing operations on buffer do not panic.

The case of too little data being read is caught by the calling function https://github.com/image-rs/image-tiff/blob/c9703dcb8a008567c5d9c6263195f56850e253a7/src/decoder/mod.rs#L1038

Comment (HeroicKatora) [reply src/decoder/mod.rs:807]: Hm, [this Stackexchange answer](https://stackoverflow.com/questions/7068183/writing-10-12-bit-tiff-files-with-libtiff-c) suggest that while libTIFF packs the bits, there is no actually standard prescribed behavior. We might have to adjust for that?

Comment (fintelia) [reply src/decoder/mod.rs:807]: Isn't any `u12` or similar data going to trigger UnsupportedColorType error below?

Comment (HeroicKatora) [reply src/decoder/mod.rs:807]: The problem is, this triggers _first_ and if the other change is reverted than we must check the units_read.

File: src/error.rs [hunk 2]
```diff
@@ -43,6 +43,14 @@ pub enum TiffFormatError {
     TiffSignatureInvalid,
     ImageFileDirectoryNotFound,
     InconsistentSizesEncountered,
+    UnexpectedCompressedData {
+        actual: usize,
+        required: usize,
```
Comment (fintelia) [inline src/error.rs:48]: Perhaps actual_bytes and required_bytes?

File: src/error.rs [hunk 3]
```diff
@@ -43,6 +43,14 @@ pub enum TiffFormatError {
...
+    },
+    InconsistentStripElements {
+        actual: usize,
+        required: usize,
```
Comment (fintelia) [inline src/error.rs:52]: Similar to above, perhaps actual_elements and required_elements?

**Review by fintelia — APPROVED** [review]

**Pull request merged**